sklearn/decomposition/incremental_pca.py

```diff
@@ -210,11 +210,18 @@ def partial_fit(self, X, y=None, check_input=True):
             self.components_ = None
 
         if self.n_components is None:
-            self.n_components_ = n_features
+            if self.components_ is None:
+                self.n_components_ = min(n_samples, n_features)
+            else:
+                self.n_components_ = self.components_.shape[0]
         elif not 1 <= self.n_components <= n_features:
             raise ValueError("n_components=%r invalid for n_features=%d, need "
                              "more rows than columns for IncrementalPCA "
                              "processing" % (self.n_components, n_features))
+        elif not self.n_components <= n_samples:
+            raise ValueError("n_components=%r must be less or equal to "
+                             "the batch number of samples "
```

Comment:
Funnily enough we were chatting with @ogrisel about this yesterday in an unrelated context. IIUC he was hoping that  `IncrementalPCA` would be able to do `partial_fit` on a small number of samples (and converge to something sensible after a few calls to partial_fit). It looks like this is not the case at the moment ...